File: abstract-factory/cpp/abstract-factory.cpp
#include <iostream>
using namespace std;

//USE ABSTRACT FACTORY WHEN YOU NEED TO WORK WITH FAMILIES OF RELATED PRODUCTS 

//Products

class Button {
    public:
        virtual ~Button() {};
        virtual void render() = 0;
};

class CheckBox {
    public:
        virtual ~CheckBox() {};
        virtual void render() = 0;
};

class LinuxButton: public Button {
    public:
        void render() {
            cout << "LinuxButton" << endl;
        }
};

class WindowsButton: public Button {
    public:
        void render() {
            cout << "WindowsButton" << endl;
        }
};

class LinuxCheckbox: public CheckBox {
    public:
        void render() {
            cout << "LinuxCheckbox" << endl;
        }
};

class WindowsCheckbox: public CheckBox {
    public:
        void render() {
            cout << "WindowsCheckbox" << endl;
        }
};

//Factories

class AbstractFactory {
    public:
        virtual ~AbstractFactory() {};
        virtual Button* createButton() = 0;
        virtual CheckBox* createCheckbox() = 0;
};

class WindowsFactory: public AbstractFactory {
    public:
        Button* createButton() {
            return new WindowsButton();
        }
        CheckBox* createCheckbox() {
            return new WindowsCheckbox();
        }
};

class LinuxFactory: public AbstractFactory {
    public:
        Button* createButton() {
            return new LinuxButton();
        }
        CheckBox* createCheckbox() {
            return new LinuxCheckbox();
        }
};

//Client code

class GUI {
    public:
        GUI(AbstractFactory *factory): factory(factory), button(NULL), checkBox(NULL) {};

        ~GUI() {
            if (this->button)
                delete this->button;
            if (this->checkBox)
                delete this->checkBox;
        }
        
        void render() {
            this->button = this->factory->createButton();
            this->checkBox = this->factory->createCheckbox();

            this->button->render();
            this->checkBox->render();
        }
    private:
        AbstractFactory *factory;
        Button* button;
        CheckBox* checkBox;
};

int main(void) {
    //linux version
    LinuxFactory linuxFactory;
    GUI guiLinux(&linuxFactory);
    guiLinux.render();

    cout << endl;

    //windows version
    WindowsFactory windowsFactory;
    GUI guiWindows(&windowsFactory);
    guiWindows.render();
    return 0;
}

File: abstract-factory/typescript/abstract-factory.ts
//USE ABSTRACT FACTORY WHEN YOU NEED TO WORK WITH FAMILIES OF RELATED PRODUCTS 

// Products

abstract class Button {
    public abstract render(): void;
}

abstract class CheckBox {
    public abstract render(): void;
}

class LinuxButton implements Button {
    public render(): void {
        console.log('LinuxButton');
    }
}

class WindowsButton implements Button {
    public render(): void {
        console.log('WindowsButton');
    }
}

class LinuxCheckbox implements CheckBox {
    public render(): void {
        console.log('LinuxCheckbox');
    }
}

class WindowsCheckBox implements CheckBox {
    public render(): void {
        console.log('WindowsCheckBox');
    }
}

// Factories

abstract class AbstractFactory {
    abstract createButton(): Button;
    abstract createCheckBox(): CheckBox;
}

class LinuxFactory implements AbstractFactory {
    public createButton(): Button {
        return new LinuxButton();
    }
    public createCheckBox(): CheckBox {
        return new LinuxCheckbox();
    }
};

class WindowsFactory implements AbstractFactory {
    public createButton(): Button {
        return new WindowsButton();
    }
    public createCheckBox(): CheckBox {
        return new WindowsCheckBox();
    }
};

//Client code

class GUI {
    private factory: AbstractFactory;
    constructor(factory: AbstractFactory) {
        this.factory = factory;
    }

    public render() {
        const button = this.factory.createButton();
        const checkBox = this.factory.createCheckBox();
        button.render();
        checkBox.render();
    }
}

const lfactory = new LinuxFactory();
const lgui = new GUI(lfactory);
lgui.render();

const wfactory = new WindowsFactory();
const wgui = new GUI(wfactory);
wgui.render();
